freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaHealthDetailsService.java

```diff
@@ -0,0 +1,132 @@
+
+    private static final Pattern NEW_NODE_PATTERN = Pattern.compile("Check connection from master to remote replica '(.[^\']*)");
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaClientFactory freeIpaClientFactory;
+
+    public HealthDetailsFreeIpaResponse getHealthDetails(String environmentCrn, String accountId) {
+        Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(environmentCrn, accountId);
+        String masterCN = findMasterCN(stack);
+        Optional<RPCResponse<Boolean>> rpcResponse = Optional.empty();
+        try {
+            rpcResponse = Optional.ofNullable(checkFreeIpaHealth(stack, masterCN));
+        } catch (FreeIpaClientException e) {
+            LOGGER.error("Unable to check the health of FreeIPA.", e);
+        }
+        return createResponse(stack, rpcResponse);
+    }
+
+    private HealthDetailsFreeIpaResponse createResponse(Stack stack, Optional<RPCResponse<Boolean>> rpcResponse) {
+        HealthDetailsFreeIpaResponse response = new HealthDetailsFreeIpaResponse();
+        response.setEnvironmentCrn(stack.getEnvironmentCrn());
+        response.setCrn(stack.getResourceCrn());
+        if (rpcResponse.isPresent()) {
+            response.setName((String) rpcResponse.get().getValue());
+            parseMessages(rpcResponse.get(), response);
+            if (isOverallHealthy(response)) {
+                response.setStatus(DetailedStackStatus.PROVISIONED.getStatus());
+            } else {
+                response.setStatus(DetailedStackStatus.UNHEALTHY.getStatus());
+            }
+        } else {
+            response.setStatus(DetailedStackStatus.UNREACHABLE.getStatus());
+            NodeHealthDetailsFreeIpaResponse nodeResponse = null;
+            nodeResponse = new NodeHealthDetailsFreeIpaResponse();
+            response.addNodeHealthDetailsFreeIpaResponses(nodeResponse);
+            nodeResponse.setStatus(Status.UNREACHABLE);
+        }
+        return response;
+    }
+
+    private String findMasterCN(Stack stack) {
+        InstanceGroup masterGroup = stack.getInstanceGroups().stream()
+                .filter(instanceGroup -> InstanceGroupType.MASTER == instanceGroup.getInstanceGroupType()).findFirst().get();
+        return masterGroup.getNotDeletedInstanceMetaDataSet().stream().findFirst().get().getDiscoveryFQDN();
+    }
+
+    private RPCResponse<Boolean> checkFreeIpaHealth(Stack stack, String masterCN) throws FreeIpaClientException {
+        FreeIpaClient freeIpaClient = freeIpaClientFactory.getFreeIpaClientForStack(stack);
+        return freeIpaClient.serverConnCheck(masterCN, masterCN);
+    }
+
+    private boolean isOverallHealthy(HealthDetailsFreeIpaResponse response) {
+        for (NodeHealthDetailsFreeIpaResponse node: response.getNodeHealthDetailsFreeIpaResponses()) {
+            if (node.getStatus().equals(Status.AVAILABLE)) {
+                return true;
+            }
+        }
+        return false;
```

Comment:
I think this doesn't do what you would expect. this would return true if 1 node is available.
I would use stream here:
`return response.getNodeHealthDetailsFreeIpaResponses().stream().allMatch(node -> Status.AVAILABLE == node.getStatus())`
or we expect it to be healthy if one node is ok?

Reply:
For HA purposes, it's still healthy if one node is ok.

Reply:
ok, cool. then `anyMatch` would do the same, but it's fine with for loop

Reply:
For HA, if there's one node able to return requests then we consider it healthy.  When we sync up the status after the HA work, we might need a new status DEGRADED, to indicate that it can still serve requests, but not everything is ok.